core/src/main/java/com/scalar/db/transaction/consensuscommit/Snapshot.java

```diff
@@ -556,11 +512,135 @@ void toSerializable(DistributedStorage storage)
+   * @throws ValidationConflictException if the scan results are changed by another transaction
+   */
+  private void validateScanResults(
+      DistributedStorage storage, Scan scan, LinkedHashMap<Key, TransactionResult> results)
+      throws ExecutionException, ValidationConflictException {
+    Scanner scanner = null;
+    try {
+      // Only get tx_id, tx_version and primary key columns because we use only them to compare
+      scan.clearProjections();
+      scan.withProjection(Attribute.ID).withProjection(Attribute.VERSION);
+      ScalarDbUtils.addProjectionsForKeys(scan, getTableMetadata(scan));
+
+      if (scan.getLimit() == 0) {
+        scanner = storage.scan(scan);
+      } else {
+        // Get a scanner without the limit if the scan has a limit
+        scanner = storage.scan(Scan.newBuilder(scan).limit(0).build());
+      }
+
+      //
+      // Initialize the iterators for the scan results and the latest scan results
+      //
+      Optional<Result> latestResult = scanner.one();
+
+      Iterator<Entry<Key, TransactionResult>> resultsIterator = results.entrySet().iterator();
+      Entry<Key, TransactionResult> resultsEntry = Iterators.getNext(resultsIterator, null);
+
+      // Compare the scan results with the latest scan results
+      while (latestResult.isPresent() && resultsEntry != null) {
+        TransactionResult latestTxResult = new TransactionResult(latestResult.get());
+        Key key = new Key(scan, latestTxResult);
+
+        if (latestTxResult.getId() != null && latestTxResult.getId().equals(id)) {
+          // The record is inserted/deleted/updated by this transaction
+
+          // Skip the record
+          latestResult = scanner.one();
+
+          if (resultsEntry.getKey().equals(key)) {
+            // The record is updated by this transaction
+
+            // Skip the record
+            resultsEntry = Iterators.getNext(resultsIterator, null);
+          } else {
+            // The record is inserted/deleted by this transaction
+          }
+
+          continue;
+        }
+
+        //
+        // Compare the records of the scan results and the latest scan results
+        //
+        if (!resultsEntry.getKey().equals(key)) {
+          // The record is inserted/deleted by another transaction
+          throwExceptionDueToAntiDependency();
+        }
+
+        if (isChanged(latestTxResult, resultsEntry.getValue())) {
+          // The record is updated by another transaction
+          throwExceptionDueToAntiDependency();
+        }
+
+        //
+        // Proceed to the next record
+        //
+        latestResult = scanner.one();
+        resultsEntry = Iterators.getNext(resultsIterator, null);
+      }
+
+      if (resultsEntry != null) {
+        // Some of the records of the scan results are deleted by another transaction
+        throwExceptionDueToAntiDependency();
+      }
+
+      if (scan.getLimit() != 0 && results.size() == scan.getLimit()) {
```

Comment:
I probably misunderstand something. But this always seems to hold when `scan.getLimit() != 0`, which is the first condition, and the latter condition is not changed after the above processing, so I'm not sure what this is checking. Can you elaborate on this?

Reply:
Thank you for the explanation offline! I understand the logic.